First step is to import the relevant pandas package and load the wiki(HTML) data.
To obtain the proper columns only we have to slice tables as tables[0]

In [84]:
import pandas as pd
import numpy as np
import requests
import json
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors

tables=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
df=pd.DataFrame(tables[0])

In [85]:
print(df)

    Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]


Next we have to select only the rows in the dataframe which have a value in the "Borough' column which is unequal to "Not assigned". The remaining rows are included in dataframe df2

In [86]:
df2=df[df['Borough'] != "Not assigned"]
print(df2)

    Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M6A        North York          Lawrence Heights
6        M6A        North York            Lawrence Manor
..       ...               ...                       ...
281      M8Z         Etobicoke  Kingsway Park South West
282      M8Z         Etobicoke                 Mimico NW
283      M8Z         Etobicoke        The Queensway West
284      M8Z         Etobicoke     Royal York South West
285      M8Z         Etobicoke            South of Bloor

[210 rows x 3 columns]


After this we have to eliminate the double items in (combined) 'Postcode' and 'Borough' columns. The neighbourhood items are combined and seperated by a ','.   

In [87]:
df3=df2.groupby(['Postcode','Borough'])['Neighbourhood'].apply(','.join).reset_index()

Next we have identify the 'Not assigned' rows in the "Neighbourhood' column. These need to be replaced with the value in the 'Borough' column. As seen below this involves only 1 item, which is replaced by the related 'Borough' field contents.

In [88]:
print(df3[df3['Neighbourhood']=='Not assigned'])

   Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Not assigned


In [89]:
df3['Neighbourhood']=df3['Neighbourhood'].str.replace("Not assigned","Queen's Park")

In [90]:
print(df3)

    Postcode      Borough                                      Neighbourhood
0        M1B  Scarborough                                      Rouge,Malvern
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough                    Guildwood,Morningside,West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
..       ...          ...                                                ...
98       M9N         York                                             Weston
99       M9P    Etobicoke                                          Westmount
100      M9R    Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...
102      M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]


Using the shape method we see that there are 103 rows remaining in the dataset after all transformations have been performed

In [91]:
df3.shape

(103, 3)

The next part of the assignment is to load the geospatial dataset and merge with the wiki dataset(df3).
First step is to load the csv and transform to dataframe.

In [92]:
geo=pd.read_csv('Geospatial_Coordinates.csv')

In [93]:
df_geo=pd.DataFrame(geo)
print(df_geo)

    Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]


The 2 dataframes(df3 and df_geo) are merged using the merge function in pandas using the Postcode/Postal Code keys. As default
we apply an inner join (default setting) 

In [94]:
df_comb=df3.merge(df_geo, left_on='Postcode', right_on='Postal Code')

After the merging of the 2 dataframes we check that we have the same number of rows left from df3 which is indeed the case (103 rows before and after merging). We drop the 'Postal Code' column as the values are the same as the 'Postcode' column.

In [95]:
df_comb=df_comb.drop(columns=['Postal Code'])
print(df_comb)


    Postcode      Borough                                      Neighbourhood  \
0        M1B  Scarborough                                      Rouge,Malvern   
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                    Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                             Woburn   
4        M1H  Scarborough                                          Cedarbrae   
..       ...          ...                                                ...   
98       M9N         York                                             Weston   
99       M9P    Etobicoke                                          Westmount   
100      M9R    Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...   
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
102      M9W    Etobicoke                                          Northwest   

      Latitude  Longitude  
0    43.806

For sake of simplification and the limitations of the Foursquare sandbox account we first investigate the Borough column in the dataset and subsequently select the boroughs with "Toronto" in their name i.e. West, East, Downtown and Central

In [96]:
df_comb['Borough'].unique()

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       "Queen's Park", 'Mississauga', 'Etobicoke'], dtype=object)

In [97]:
df_comb=df_comb[(df_comb.Borough == 'West Toronto') | (df_comb.Borough == 'East Toronto') | (df_comb.Borough == 'Downtown Toronto') | (df_comb.Borough == 'Central Toronto')].reset_index() 
print(df_comb)

    index Postcode           Borough  \
0      37      M4E      East Toronto   
1      41      M4K      East Toronto   
2      42      M4L      East Toronto   
3      43      M4M      East Toronto   
4      44      M4N   Central Toronto   
5      45      M4P   Central Toronto   
6      46      M4R   Central Toronto   
7      47      M4S   Central Toronto   
8      48      M4T   Central Toronto   
9      49      M4V   Central Toronto   
10     50      M4W  Downtown Toronto   
11     51      M4X  Downtown Toronto   
12     52      M4Y  Downtown Toronto   
13     53      M5A  Downtown Toronto   
14     54      M5B  Downtown Toronto   
15     55      M5C  Downtown Toronto   
16     56      M5E  Downtown Toronto   
17     57      M5G  Downtown Toronto   
18     58      M5H  Downtown Toronto   
19     59      M5J  Downtown Toronto   
20     60      M5K  Downtown Toronto   
21     61      M5L  Downtown Toronto   
22     63      M5N   Central Toronto   
23     64      M5P   Central Toronto   


We select the relevant packages/libraries necessary for the map viz and clustering.

In [98]:
import folium
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans

Let's apply one of the data points as starting point

In [99]:
latitude=43.806686
longitude=-79.194353


Next we are going to cycle through all datapoints and plot them on the folium map in color blue for now

In [100]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_comb.Latitude, df_comb.Longitude, df_comb.Borough, df_comb.Neighbourhood):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next we include the Foursquare identification to apply the Foursquare API

In [101]:
LIMIT = 100
CLIENT_ID = 'GHNHGWRHFKZE5DGOGHZUYBEDFWXFDMAUPW5JVVATOK3YC0ZH' # your Foursquare ID
CLIENT_SECRET = 'YBL2RZCIKT3F5A2KOWQ0ET04D0YII0XE5GRTSLFJOC1UCHAY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: GHNHGWRHFKZE5DGOGHZUYBEDFWXFDMAUPW5JVVATOK3YC0ZH
CLIENT_SECRET:YBL2RZCIKT3F5A2KOWQ0ET04D0YII0XE5GRTSLFJOC1UCHAY


Let's obtain the nearby venues within a 500 radius (again we want to limit this while we are using a Foursquare sandbox account) and define the columns we want in our dataset 

In [102]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

We are going to apply above function on each neighboorhood and create data frame toronto_venues

In [103]:
toronto_venues = getNearbyVenues(names=df_comb['Neighbourhood'],
                                   latitudes=df_comb['Latitude'],
                                   longitudes=df_comb['Longitude']
                                  )
toronto_venues=pd.DataFrame(toronto_venues)

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

Show the size of the dataframe

In [104]:
print(toronto_venues.shape)
toronto_venues.head()

(1702, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   The Beaches               43.676357               -79.293031   
1   The Beaches               43.676357               -79.293031   
2   The Beaches               43.676357               -79.293031   
3   The Beaches               43.676357               -79.293031   
4   The Beaches               43.676357               -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

Display the number of venues for each neighboorhood

In [105]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       16   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      15   
Cabbagetown,St. James Town                                              47   
Central Bay Street                                                      82   
Chinatown,Grange Park,Kensington Market                                 97   
Christie                                                                17   
Church and Wellesley                                                    87   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              37   
Davisville North                                                        10   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      15   
Design Exchange,Toronto Dominion Centre                                100   
Dovercourt Village,Dufferin                                             16   
First Canadian Place,Underground city                                  100   
Forest Hill North,Forest Hill West                                       4   
Harbord,University of Toronto                                           35   
Harbourfront                                                            49   
Harbourfront East,Toronto Islands,Union Station                        100   
High Park,The Junction South                                            24   
Lawrence Park                                                            3   
Little Portugal,Trinity                                                 64   
Moore Park,Summerhill East                                               3   
North Toronto West                                                      21   
Parkdale,Roncesvalles                                                   15   
Rosedale                                                                 4   
Roselawn                                                                 3   
Runnymede,Swansea                                                       37   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         99   
Studio District                                                         37   
The Annex,North Midtown,Yorkville                                       22   
The Beaches                                                              5   
The Beaches West,India Bazaar                                           19   
The Danforth West,Riverdale                                             42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Berczy Park                                                              55   
Brockton,Exhibition Place,Parkdale Village                               22   
Business Reply Mail Processing Centre 969 Eastern                        16   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       15   
Cabbagetown,St. James Town                                               47   
Central Bay Street                                                       82   
Chinatown,Grange Park,Kensington Market                                  97   
Christie                                                                 17   
Church and Wellesley                                          

Lets find out how many unique venue categories are included

In [106]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 230 uniques categories.


We use one hot encoding to create dummy variables, showing the unique venue categories in columns showing either 0(not applicable) or 1(applicable)

In [107]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0   The Beaches                  0        0                   0             0   
1   The Beaches                  0        0                   0             0   
2   The Beaches                  0        0                   0             0   
3   The Beaches                  0        0                   0             0   
4   The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0             0  ...      1              0                              0   
1             0  ...      0              0                              0   
2             0  ...      0              0                              0   
3             0  ...      0              0                              0   
4             0  ...      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  \
0                 0                      0         0          0            0   
1                 0                      0         0          0            0   
2                 0                      0         0          0            0   
3                 0                      0         0          0            0   
4                 0                      0         0          0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 231 columns]

let's see the size of the dateframe

In [108]:
toronto_onehot.shape

(1702, 231)

To prepare for clustering we want to group by neighborhood and obtain the mean of the frequency of occurring of each category 

In [115]:
toronto_grouped=toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011494   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19                                       Harbourfront           0.000000   
20    Harbourfront East,Toronto Islands,Union Station           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                           Rosedale           0.000000   
28                                           Roselawn           0.000000   
29                                  Runnymede,Swansea           0.000000   
30                            Ryerson,Garden District           0.000000   
31                                     St. James Town           0.000000   
32                    Stn A PO Boxes 25 The Esplanade           0.000000   
33                                    Studio District           0.000000   
34                  The Annex,North Midtown,Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                      The Beaches West,India Bazaar           0.000000   
37                        The Danforth West,Riverdale           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.066667            0.066667      0.066667        0.133333   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000       

Let's display the shape of the new dataframe

In [116]:
toronto_grouped.shape

(38, 231)

next we want to know the top 5 of venues per neighbourhood

In [119]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1              Bakery  0.05
2        Cocktail Bar  0.04
3      Farmers Market  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
                venue  freq
0      Breakfast Spot  0.09
1                Café  0.09
2         Coffee Shop  0.09
3              Bakery  0.05
4  Italian Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0         Yoga Studio  0.06
1                 Spa  0.06
2       Garden Center  0.06
3              Garden  0.06
4  Light Rail Station  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.13
1   Airport Servi

Now we are going to put this in a dataframe, displaying the outcome in descending order

In [126]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [129]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1                Bakery           Cheese Shop            Steakhouse   
2                  Café           Coffee Shop          Climbing Gym   
3         Auto Workshop         Garden Center                Garden   
4       Airport Service      Airport Terminal       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot      Asian Restaurant        Cosmetics Shop   
1    Seafood Restaurant                  Café        Farmers Market   
2     Convenience Store         Burrito Place               Stadium   
3  Fast Food Restaurant    Light Rail Station        Farmers Market   
4                   Bar           Coffee Shop      Sculpture Garden   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel            Restaurant    American Restaurant  
1              Beer Bar          Cocktail Bar     Italian Restaurant  
2  Caribbean Restaurant    Italian Restaurant             Restaurant  
3            Comic Shop                  Park             Restaurant  
4              Boutique         Boat or Ferry           Airport Gate

We use k-means to cluster the neighbourhood 

In [130]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a dataframe that includes the cluster as well as the top 10 venues for each neighboorhood

In [131]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_comb

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() 

index Postcode          Borough                  Neighbourhood   Latitude  \
0     37      M4E     East Toronto                    The Beaches  43.676357   
1     41      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2     42      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3     43      M4M     East Toronto                Studio District  43.659526   
4     44      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4          Neighborhood                   Pub   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park            Board Shop   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               3                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                  Park     Health Food Store                   Trail   
1        Ice Cream Shop    Italian Restaurant  Furniture / Home Store   
2                   Pub               Brewery            Burger Joint   
3                Bakery    Italian Restaurant     American Restaurant   
4              Bus Line           Yoga Studio                   Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1          Liquor Store                   Spa             Juice Bar   
2         Burrito Place  Fast Food Restaurant        Sandwich Place   
3        Sandwich Place      Stationery Store           Fish Market   
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

  9th Most Common Venue       10th Most Common Venue  
0               Dog Run             Doner Restaurant  
1             Bookstore                      Brewery  
2     Fish & Chips Shop                   Steakhouse  
3       Coworking Space    Latin American Restaurant  
4     Electronics Store  Eastern European Restaurant

As last step we are showing the map with the clusters

In [132]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters